File: binding/src/libqprogressbar.cpp
#include <QEvent>
#include <QPaintEvent>
#include <QProgressBar>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyleOptionProgressBar>
#include <QWidget>
#include <qprogressbar.h>
#include "libqprogressbar.h"
#include "libqprogressbar.hxx"

QProgressBar* QProgressBar_new(QWidget* parent) {
	 return new VirtualQProgressBar(parent);
}

QProgressBar* QProgressBar_new2() {
	 return new VirtualQProgressBar();
}

libqt_string QProgressBar_Tr(const char* s) {
	QString _ret = QProgressBar::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QProgressBar_Minimum(const QProgressBar* self) {
	return self->minimum();
}

int QProgressBar_Maximum(const QProgressBar* self) {
	return self->maximum();
}

int QProgressBar_Value(const QProgressBar* self) {
	return self->value();
}

libqt_string QProgressBar_Text(const QProgressBar* self) {
	QString _ret = self->text();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QProgressBar_SetTextVisible(QProgressBar* self, bool visible) {
	self->setTextVisible(visible);
}

bool QProgressBar_IsTextVisible(const QProgressBar* self) {
	return self->isTextVisible();
}

int QProgressBar_Alignment(const QProgressBar* self) {
	return self->alignment();
}

void QProgressBar_SetAlignment(QProgressBar* self, int alignment) {
	self->setAlignment(static_cast<QFlags<Qt::AlignmentFlag>>(alignment));
}

QSize* QProgressBar_SizeHint(const QProgressBar* self) {
	return new QSize(self->sizeHint());
}

QSize* QProgressBar_MinimumSizeHint(const QProgressBar* self) {
	return new QSize(self->minimumSizeHint());
}

int QProgressBar_Orientation(const QProgressBar* self) {
	return self->orientation();
}

void QProgressBar_SetInvertedAppearance(QProgressBar* self, bool invert) {
	self->setInvertedAppearance(invert);
}

bool QProgressBar_InvertedAppearance(const QProgressBar* self) {
	return self->invertedAppearance();
}

void QProgressBar_SetTextDirection(QProgressBar* self, int textDirection) {
	self->setTextDirection(static_cast<QProgressBar::Direction>(textDirection));
}

int QProgressBar_TextDirection(const QProgressBar* self) {
	return self->textDirection();
}

void QProgressBar_SetFormat(QProgressBar* self, const libqt_string format) {
	self->setFormat(QString::fromUtf8(format.data, format.len));
}

void QProgressBar_ResetFormat(QProgressBar* self) {
	self->resetFormat();
}

libqt_string QProgressBar_Format(const QProgressBar* self) {
	QString _ret = self->format();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QProgressBar_Reset(QProgressBar* self) {
	self->reset();
}

void QProgressBar_SetRange(QProgressBar* self, int minimum, int maximum) {
	self->setRange(minimum, maximum);
}

void QProgressBar_SetMinimum(QProgressBar* self, int minimum) {
	self->setMinimum(minimum);
}

void QProgressBar_SetMaximum(QProgressBar* self, int maximum) {
	self->setMaximum(maximum);
}

void QProgressBar_SetValue(QProgressBar* self, int value) {
	self->setValue(value);
}

void QProgressBar_SetOrientation(QProgressBar* self, int orientation) {
	self->setOrientation(static_cast<Qt::Orientation>(orientation));
}

void QProgressBar_ValueChanged(QProgressBar* self, int value) {
	self->valueChanged(value);
}

void QProgressBar_Connect_ValueChanged(QProgressBar* self, intptr_t slot) {
    void (*slotFunc)(QProgressBar*, int) = reinterpret_cast<void (*)(QProgressBar*, int)>(slot);
    QProgressBar::connect(self, &QProgressBar::valueChanged, [self, slotFunc](int value) {
	slotFunc(self, value);
    });
}

libqt_string QProgressBar_Tr2(const char* s, const char* c) {
	QString _ret = QProgressBar::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QProgressBar_Tr3(const char* s, const char* c, int n) {
	QString _ret = QProgressBar::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
libqt_string QProgressBar_QBaseText(const QProgressBar* self) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_Text_IsBase(true);
	QString _ret = vqprogressbar->text();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QProgressBar_OnText(const QProgressBar* self, intptr_t slot) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_Text_Callback(reinterpret_cast<VirtualQProgressBar::QProgressBar_Text_Callback>(slot));
}
}

// Base class handler implementation
QSize* QProgressBar_QBaseSizeHint(const QProgressBar* self) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_SizeHint_IsBase(true);
	return new QSize(vqprogressbar->sizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QProgressBar_OnSizeHint(const QProgressBar* self, intptr_t slot) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_SizeHint_Callback(reinterpret_cast<VirtualQProgressBar::QProgressBar_SizeHint_Callback>(slot));
}
}

// Base class handler implementation
QSize* QProgressBar_QBaseMinimumSizeHint(const QProgressBar* self) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_MinimumSizeHint_IsBase(true);
	return new QSize(vqprogressbar->minimumSizeHint());
}
}

// Auxiliary method to allow providing re-implementation
void QProgressBar_OnMinimumSizeHint(const QProgressBar* self, intptr_t slot) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_MinimumSizeHint_Callback(reinterpret_cast<VirtualQProgressBar::QProgressBar_MinimumSizeHint_Callback>(slot));
}
}

// Derived class handler implementation
bool QProgressBar_Event(QProgressBar* self, QEvent* e) {
	auto* vqprogressbar = dynamic_cast<VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
	return vqprogressbar->event(e);
	} else {
	return self->QProgressBar::event(e);
}
}

// Base class handler implementation
bool QProgressBar_QBaseEvent(QProgressBar* self, QEvent* e) {
	auto* vqprogressbar = dynamic_cast<VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_Event_IsBase(true);
	return vqprogressbar->event(e);
}
}

// Auxiliary method to allow providing re-implementation
void QProgressBar_OnEvent(QProgressBar* self, intptr_t slot) {
	auto* vqprogressbar = dynamic_cast<VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_Event_Callback(reinterpret_cast<VirtualQProgressBar::QProgressBar_Event_Callback>(slot));
}
}

// Derived class handler implementation
void QProgressBar_PaintEvent(QProgressBar* self, QPaintEvent* param1) {
	auto* vqprogressbar = dynamic_cast<VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
	vqprogressbar->paintEvent(param1);
	} else {
	self->QProgressBar::paintEvent(param1);
}
}

// Base class handler implementation
void QProgressBar_QBasePaintEvent(QProgressBar* self, QPaintEvent* param1) {
	auto* vqprogressbar = dynamic_cast<VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_PaintEvent_IsBase(true);
	vqprogressbar->paintEvent(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QProgressBar_OnPaintEvent(QProgressBar* self, intptr_t slot) {
	auto* vqprogressbar = dynamic_cast<VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_PaintEvent_Callback(reinterpret_cast<VirtualQProgressBar::QProgressBar_PaintEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QProgressBar_InitStyleOption(const QProgressBar* self, QStyleOptionProgressBar* option) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
	vqprogressbar->initStyleOption(option);
	} else {
	self->QProgressBar::initStyleOption(option);
}
}

// Base class handler implementation
void QProgressBar_QBaseInitStyleOption(const QProgressBar* self, QStyleOptionProgressBar* option) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_InitStyleOption_IsBase(true);
	vqprogressbar->initStyleOption(option);
}
}

// Auxiliary method to allow providing re-implementation
void QProgressBar_OnInitStyleOption(const QProgressBar* self, intptr_t slot) {
	auto* vqprogressbar = dynamic_cast<const VirtualQProgressBar*>(self);
	if (vqprogressbar && vqprogressbar->isVirtualQProgressBar) {
vqprogressbar->setQProgressBar_InitStyleOption_Callback(reinterpret_cast<VirtualQProgressBar::QProgressBar_InitStyleOption_Callback>(slot));
}
}

void QProgressBar_Delete(QProgressBar* self) {
    delete self;
}

File: binding/include/libqprogressbar.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qprogressbar.html
pub const qprogressbar = struct {

    /// New constructs a new QProgressBar object.
    pub fn New(parent: ?*anyopaque) QtC.QProgressBar {
        return qtc.QProgressBar_new(@ptrCast(parent));
    }


    /// New2 constructs a new QProgressBar object.
    pub fn New2() QtC.QProgressBar {
        return qtc.QProgressBar_new2();
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QProgressBar_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprogressbar.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Minimum(self: ?*anyopaque, ) i32 {
        return qtc.QProgressBar_Minimum(@ptrCast(self));
    }

    pub fn Maximum(self: ?*anyopaque, ) i32 {
        return qtc.QProgressBar_Maximum(@ptrCast(self));
    }

    pub fn Value(self: ?*anyopaque, ) i32 {
        return qtc.QProgressBar_Value(@ptrCast(self));
    }

    pub fn Text(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QProgressBar_Text(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprogressbar.Text: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnText(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QProgressBar_OnText(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseText(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QProgressBar_QBaseText(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprogressbar.Text: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetTextVisible(self: ?*anyopaque, visible: bool) void {
        qtc.QProgressBar_SetTextVisible(@ptrCast(self), visible);
    }

    pub fn IsTextVisible(self: ?*anyopaque, ) bool {
        return qtc.QProgressBar_IsTextVisible(@ptrCast(self));
    }

    pub fn Alignment(self: ?*anyopaque, ) i32 {
        return qtc.QProgressBar_Alignment(@ptrCast(self));
    }

    pub fn SetAlignment(self: ?*anyopaque, alignment: i32) void {
        qtc.QProgressBar_SetAlignment(@ptrCast(self), @intCast(alignment));
    }

    pub fn SizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QProgressBar_SizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QProgressBar_OnSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QProgressBar_QBaseSizeHint(@ptrCast(self));
    }

    pub fn MinimumSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QProgressBar_MinimumSizeHint(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnMinimumSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QProgressBar_OnMinimumSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMinimumSizeHint(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QProgressBar_QBaseMinimumSizeHint(@ptrCast(self));
    }

    pub fn Orientation(self: ?*anyopaque, ) i32 {
        return qtc.QProgressBar_Orientation(@ptrCast(self));
    }

    pub fn SetInvertedAppearance(self: ?*anyopaque, invert: bool) void {
        qtc.QProgressBar_SetInvertedAppearance(@ptrCast(self), invert);
    }

    pub fn InvertedAppearance(self: ?*anyopaque, ) bool {
        return qtc.QProgressBar_InvertedAppearance(@ptrCast(self));
    }

    pub fn SetTextDirection(self: ?*anyopaque, textDirection: i32) void {
        qtc.QProgressBar_SetTextDirection(@ptrCast(self), @intCast(textDirection));
    }

    pub fn TextDirection(self: ?*anyopaque, ) i32 {
        return qtc.QProgressBar_TextDirection(@ptrCast(self));
    }

    pub fn SetFormat(self: ?*anyopaque, format: []const u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QProgressBar_SetFormat(@ptrCast(self), format_str);
    }

    pub fn ResetFormat(self: ?*anyopaque, ) void {
        qtc.QProgressBar_ResetFormat(@ptrCast(self));
    }

    pub fn Format(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QProgressBar_Format(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprogressbar.Format: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Reset(self: ?*anyopaque, ) void {
        qtc.QProgressBar_Reset(@ptrCast(self));
    }

    pub fn SetRange(self: ?*anyopaque, minimum: i32, maximum: i32) void {
        qtc.QProgressBar_SetRange(@ptrCast(self), minimum, maximum);
    }

    pub fn SetMinimum(self: ?*anyopaque, minimum: i32) void {
        qtc.QProgressBar_SetMinimum(@ptrCast(self), minimum);
    }

    pub fn SetMaximum(self: ?*anyopaque, maximum: i32) void {
        qtc.QProgressBar_SetMaximum(@ptrCast(self), maximum);
    }

    pub fn SetValue(self: ?*anyopaque, value: i32) void {
        qtc.QProgressBar_SetValue(@ptrCast(self), value);
    }

    pub fn SetOrientation(self: ?*anyopaque, orientation: i32) void {
        qtc.QProgressBar_SetOrientation(@ptrCast(self), @intCast(orientation));
    }

    pub fn ValueChanged(self: ?*anyopaque, value: i32) void {
        qtc.QProgressBar_ValueChanged(@ptrCast(self), value);
    }

    pub fn OnValueChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QProgressBar_Connect_ValueChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Event(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QProgressBar_Event(@ptrCast(self), @ptrCast(e));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QProgressBar_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QProgressBar_QBaseEvent(@ptrCast(self), @ptrCast(e));
    }

    pub fn PaintEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QProgressBar_PaintEvent(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QProgressBar_OnPaintEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QProgressBar_QBasePaintEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn InitStyleOption(self: ?*anyopaque, option: ?*anyopaque) void {
        qtc.QProgressBar_InitStyleOption(@ptrCast(self), @ptrCast(option));
    }

    /// Allows for overriding the related default method
    pub fn OnInitStyleOption(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QProgressBar_OnInitStyleOption(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseInitStyleOption(self: ?*anyopaque, option: ?*anyopaque) void {
        qtc.QProgressBar_QBaseInitStyleOption(@ptrCast(self), @ptrCast(option));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QProgressBar_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprogressbar.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QProgressBar_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprogressbar.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QProgressBar_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/progressbar.html#types
pub const enums = struct {
    pub const Direction = enum {
        pub const TopToBottom: i32 = 0;
        pub const BottomToTop: i32 = 1;
    };

};
